plugins/woocommerce-blocks/assets/js/blocks/product-elements/add-to-cart-form/components/upgrade-notice/index.tsx

```diff
@@ -0,0 +1,95 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { Notice, Button } from '@wordpress/components';
+import { createInterpolateElement } from '@wordpress/element';
+import { recordEvent } from '@woocommerce/tracks';
+import { dispatch, select } from '@wordpress/data';
+import { findBlock } from '@woocommerce/utils';
+import {
+	createBlock,
+	// @ts-expect-error Type definitions for this function are missing in Gutenberg
+	createBlocksFromInnerBlocksTemplate,
+} from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import metadata from '../../block.json';
+import getInnerBlocksTemplate from '../../../../add-to-cart-with-options/utils/get-inner-blocks-template';
+
+const upgradeToBlockifiedAddtoCartWithOptions = async (
+	blockClientId: string
+) => {
+	const blocks = select( 'core/block-editor' ).getBlocks();
+	const foundBlock = findBlock( {
+		blocks,
+		findCondition: ( block ) =>
+			block.name === metadata.name && block.clientId === blockClientId,
+	} );
+
+	if ( ! foundBlock ) {
+		return;
+	}
+
+	const newBlock = createBlock(
+		'woocommerce/add-to-cart-with-options',
+		{
+			isDescendentOfSingleProductBlock:
+				foundBlock.attributes.isDescendentOfSingleProductBlock,
+		},
+		createBlocksFromInnerBlocksTemplate(
+			getInnerBlocksTemplate(
+				foundBlock.attributes.quantitySelectorStyle
+			)
+		)
+	);
+	dispatch( 'core/block-editor' ).replaceBlock(
+		foundBlock.clientId,
+		newBlock
+	);
+};
+
+export const UpgradeNotice = ( {
+	blockClientId,
+}: {
+	blockClientId: string;
+} ) => {
+	const notice = createInterpolateElement(
+		__(
+			'Upgrade the Add to Cart with Options block to <strongText /> for more features!',
+			'woocommerce'
+		),
+		{
+			strongText: (
+				<strong>
+					{ __( `a new blockified experience`, 'woocommerce' ) }
+				</strong>
+			),
+		}
+	);
+
+	const buttonLabel = __(
+		'Upgrade to the blockified Add to Cart with Options block',
+		'woocommerce'
+	);
+
+	const handleClick = () => {
+		upgradeToBlockifiedAddtoCartWithOptions( blockClientId );
+		recordEvent( 'blocks_add_to_cart_with_options_migration', {
+			transform_to: 'blockified',
+		} );
+	};
```

Comment:
After recent troubles we had with `createBlocksFromInnerBlocksTemplate` in various areas (https://github.com/woocommerce/woocommerce/issues/53765), I wanted to check how it behaves in Widgets and Patterns.

It seems the upgrade doesn't happen in widgets in classic theme and I think it's ok, because Single Product block doesn't work there anyway.

However, the event is recorded anyway. Considering the `upgradeToBlockifiedAddtoCartWithOptions` may not pass due to:

```
    if ( ! foundBlock ) {
		return;
	}
```

I think the `upgradeToBlockifiedAddtoCartWithOptions` should return boolean and the track should be recorded or not depends on the outcome:

```suggestion
	const handleClick = () => {
		const upgraded = upgradeToBlockifiedAddtoCartWithOptions( blockClientId );
		if ( upgraded ) {
			recordEvent( 'blocks_add_to_cart_with_options_migration', {
			    transform_to: 'blockified',
		    } );
		}
	};
```

It's minor and rather unlikely scenario but feels cleaner. The same would apply to downgrade.

Reply:
Good catch. We should probably also hide the block in the widgets screen. :grimacing: I will take a look at it.

For now, I implemented your suggestion in 50ce7c1c8834df7c32af375464da9f2812bfe5e2.

Reply:
> Good catch. We should probably also hide the block in the widgets screen. 😬 I will take a look at it.

FWIW it can be hidden in here (wouldn't be surprised if you knew this already 😅 ):

https://github.com/woocommerce/woocommerce/blob/62b81a321a219f8e241e7f12ca9394fbd3aa9e97/plugins/woocommerce/src/Blocks/BlockTypesController.php#L521C6-L536C4

Reply:
Thanks! I opened a PR here: https://github.com/woocommerce/woocommerce/pull/54087. Apparently switching from `registerBlockType()` to `registerProductBlockType()` was enough. Not sure if we should add the block to the PHP list as well, but considering we have https://github.com/woocommerce/woocommerce/issues/53691, I don't think it's needed. :thinking:

Reply:
> considering we have https://github.com/woocommerce/woocommerce/issues/53691, I don't think it's needed. 🤔

Actually, I even picked it up today and PR is nearly ready (https://github.com/woocommerce/woocommerce/pull/54086) so definitely not worth the effort!

Reply:
Awesome! :heart_eyes: Feel free to ping me if you need a review!